---
title: "clusterProfiler"
author: "Tim Vigers"
date: "today"
date-format: long
format: revealjs
self-contained: true
editor: source
bibliography: /Users/timvigers/Library/CloudStorage/Dropbox/Miscellaneous/zotero.bib
csl: american-medical-association.csl
---

```{r include=FALSE}
library(knitr)
```

## Functional enrichment

- Methods for interpreting results of various high-throughput omics studies. 
- Most omics analyses result in a list of genes (or proteins, metabolites, etc. that are associated with a gene) that are differentially expressed.
  - For example, comparing protein levels between two groups.
- How do we understand a list of differentially expressed genes in biological context?

## Functional enrichment terminology

- Gene set: an unordered collection of functionally related genes (a pathway) @yuBiomedicalKnowledgeMining
- Gene ontology (GO): a formal representation of three aspects of biological knowledge @GeneOntologyOverview
  - Molecular function
    - E.g., "catalysis" or "transport"
  - Cellular component
    - Either cellular compartments (e.g. "mitochondrion") or stable macromolecular complexes (e.g. "ribosome")

## Functional enrichment terminology

- Gene ontology (continued):
  - Biological processes
    - Larger processes made up of multiple molecular functions. 
    - E.g., "signal transduction" or "glucose membrane transport"
    - Not necessarily equivalent to a pathway

## GO graph

- Organized as a directed acyclic graph (DAG)
- In general, child terms are more specific than parent terms
  - E.g. *biosynthetic process* is a subtype of *metabolic process* and a *hexose* is a subtype of *monosaccharide*.
  
## GO graph example

![](hexose-biosynthetic-process.png)

## KEGG

- Kyoto Encyclopedia of Genes and Genomes
- A "manually curated database resource integrating various biological objects categorized into systems, genomic, chemical and health information." @kanehisaKEGGTaxonomybasedAnalysis2023
- Sixteen databases in four broad categories:
  1. Systems information (includes the pathway database)
  2. Genomic information
  3. Chemical information
  4. Health information

##

![](KEGG_catgories.png)

[KEGG Gluconeogenesis Pathway](https://www.genome.jp/pathway/map00010)

## More categories!

KEGG pathways can be further subdivided into:

  1. Metabolism 
  2. Genetic Information Processing 
  3. Environmental Information Processing 
  4. Cellular Processes 
  5. Organismal Systems 
  6. Human Diseases 
  7. Drug Development

## Other gene sets

- GO, KEGG, and Reactome @gillespieReactomePathwayKnowledgebase2022 are the most frequently used @yuBiomedicalKnowledgeMining
- Can use any pathway database for these analyses
- Alternatives include: 
  - Disease Ontology (DO)
  - Disease Gene Network (DisGeNET)
  - wikiPathways
  - Molecular Signatures Database (MSigDb)
- Unfortunately, the pathway database can have a significant impact on results
  
## Over representation analysis (ORA)

- Takes a list of differentially expressed genes and tests whether genes from various pathways are present in the list more often than expected. 
- Usually a hypergeometric test @boyleGOTermFinderOpen2004

## Over representation analysis (ORA)

$$
p = 1 - \frac{\sum_{i=0}^{k-1}{M\choose i}{N-M \choose n-i}}{N \choose n}
$$

- $N$ is the number of genes in the background distribution (usually all annotated genes)
- $n$ is the number of genes of interest
- $M$ is the number of genes annotated to the particular gene set (pathway) $S$
- $k$ is the number of genes of interest that are annotated to $S$

## ORA example

- A background of 10,000 genes, of which 260 are categorized as "axon guidance" ($S$). 
- We find that 1000 genes are differentially expressed, and 50 of those are categorized as "axon guidance."
- Is this statistically significant over representation?

## ORA example

```{r echo=TRUE}
phyper(q = 50-1,m = 260,n=10000-260,k=1000,lower.tail = F)
m = matrix(c(50,1000-50,260-50,10000-1000-260+50),nr=2)
colnames(m) = c("DE","Not DE")
rownames(m) = c("In S","Not in S")
kable(m)
fisher.test(m,alternative = "g")$p.value
```

## ORA problems

- Does not automatically account for the direction of differential expression
  - Can look at down-regulated and up-regulated genes separately. 
- Also does not take effect size into account
  - Will miss small but coordinated changes across lots of genes
    - These are likely more biologically relevant.
  
## Gene set enrichment analysis (GSEA)

- Basically, the goal is to test whether members of a gene set $S$ are distributed randomly throughout a gene list $L$. @subramanianGeneSetEnrichment2005
- An enrichment score (ES) is calculated by: 
  1. Walking down the gene list $L$ (usually ranked by effect size/correlation with the phenotype). 
  2. When a gene is in set $S$, the ES increases, and decreases when not in $S$.
      - Increment depends on the statistic in $L$

## Gene set enrichment analysis (GSEA)

  3. The final ES is the maximum deviation from 0, and corresponds to a weighted Kolmogorovâ€“Smirnov-like statistic. @subramanianGeneSetEnrichment2005
  4. The statistical significance of the ES is estimated using an empirical phenotype-based permutation test. Shuffling the phenotype preserves gene-gene correlations, and is better than shuffling gene labels.
  
## clusterProfiler

- Essentially, a set of wrapper functions that simplify functional enrichment analyses.
- Provides really nice visualizations for pathway analyses.
- Also includes utility functions that simplify conversion between gene identifiers (e.g. from gene symbol to Entrez ID).
- Allows for comparison of multiple gene lists.
- Uses `dplyr` verbs (`mutate()`, etc.) for altering figures and results.

## Reactome ORA example

```{r echo=TRUE}
library(clusterProfiler)
library(ReactomePA)
data(geneList, package="DOSE")
head(geneList)
de <- names(geneList)[abs(geneList) > 1.5]
head(de)
```

## Reactome ORA example

```{r echo=TRUE}
x <- enrichPathway(gene=de)
res = x@result
kable(head(res[,1:6]),row.names = F)
```

## Reactome ORA barplot

```{r echo=TRUE}
library(enrichplot)
barplot(x, showCategory=10) 
```

## Reactome ORA dotplot

```{r echo=TRUE}
dotplot(x, showCategory=10) 
```

## Reactome ORA gene-concept network

```{r echo=TRUE}
library(DOSE)
# Convert gene ID to Symbol
edox <- setReadable(x, 'org.Hs.eg.db', 'ENTREZID')
# Network plot
cnetplot(edox, foldChange=geneList)
# Circular plot
cnetplot(edox, foldChange=geneList, circular = TRUE)
```

## Reactome ORA gene-concept network

```{r}
cnetplot(edox, foldChange=geneList, circular = TRUE)
```

## Viewing a specific pathway

```{r echo=TRUE}
viewPathway("E2F mediated regulation of DNA replication", 
            readable = TRUE, 
            foldChange = geneList)
```

## Reactome ORA heatmap

```{r echo=TRUE}
heatplot(edox, foldChange=geneList, showCategory=10)
```

## Reactome ORA treeplot

Treeplots give you a slightly higher level overview by clustering terms based on similarity.

```{r}
edox <- pairwise_termsim(edox)
treeplot(edox)
```

## KEGG GSEA

- All of the previous plots can be used with either ORA or GSEA. I would recommend using GSEA results whenever possible.
- There are some GSEA-specific plots that can be useful as well.

## KEGG GSEA ridgeplot

- The ridgeplot helps users to interpret up vs. down-regulated pathways.

```{r echo=TRUE}
x <- gseKEGG(geneList)
edox <- setReadable(x, 'org.Hs.eg.db', 'ENTREZID')
edox = pairwise_termsim(edox)
```

## KEGG GSEA ridgeplot

```{r echo=TRUE}
ridgeplot(edox)
```

## KEGG GSEA statistic

```{r echo=TRUE}
gseaplot2(edox, geneSetID = 1:3, pvalue_table = TRUE,
          color = c("#E495A5", "#86B875", "#7DB0DD"), ES_geom = "dot")
```

## KEGG GSEA PubMed search

- Can plot the number and proportion of publications containing specific terms over time.

```{r echo=TRUE}
terms <- edox$Description[1:5]
pmcplot(terms, 2010:2020, proportion=FALSE)
```

## Theme comparison

- The `compareCluster` function allows you to perform enrichment analysis on multiple gene lists at once.
- There is a formula interface that allows for somewhat complex comparisons.

```{r echo=TRUE}
df <- data.frame(Entrez=names(geneList), FC=geneList)
df$group <- "upregulated"
df$group[df$FC < 0] <- "downregulated"
df$othergroup <- "A"
df$othergroup[abs(df$FC) > 2] <- "B"
```

## Theme comparison

```{r echo=TRUE}
kable(head(df))
```

## Theme comparison

```{r echo=TRUE}
formula_res <- compareCluster(Entrez~group+othergroup, data=df, fun="enrichKEGG")
dotplot(formula_res, x="group") + facet_grid(~othergroup)
```

## Theme comparison

```{r}
cnetplot(formula_res)
```

## Theme comparison

```{r echo=TRUE}
data(gcSample)
str(gcSample) 
ck <- compareCluster(geneCluster = gcSample, fun = enrichKEGG)
ck <- setReadable(ck, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
```

## Theme comparison

```{r}
cnetplot(ck)
```

## Questions?

## References
